use fetch arrayBuffer() if present

d4l3k opened the pull request:
This PR makes npyjs use `Response.arrayBuffer()` if present since it's way way faster than using a FileReader in Firefox. 

From my tests with a 4.8MB file it now takes about ~448ms instead of 50 seconds.

For backwards compatibility this keeps the FileReader fallback but I don't think it's actually needed in any modern browsers. It also cleans up the error handling to actually throw errors since that's generally what people expect to happen if a file fails to load.

Test plan:

```
mocha
```

j6k4m8 commented:
Holy smokes, HUGE performance improvement!

I would indeed love to get mocha running on this repository; did your comment mean you wanted to add to this PR before merging?  Otherwise, I'm happy to merge now :) Thank you a ton, @d4l3k!

d4l3k commented:
I wasn't sure why it was using the FileReader API in the first place. Is it fine to delete that? 

Also using async await instead of promises would clean this up a lot but I wasn't sure what versions you were targeting.

But it's probably fine to land as is, just not as clean as it could be

j6k4m8 commented:
I think fine to delete; doesn't feel like we'd lose any functionality, right? 

When I wrote this originally, I couldn't part with promises yet; I think that it is probably time now, and if that was an offer to refactor, I would gladly accept 😍

d4l3k commented:
Just updated it to use async/await and get rid of the FileReader usage. Also updated to use a http server in the test so it actually tests the library E2E.

j6k4m8 commented:
This is awesome, good to merge @d4l3k?? Great stuff!

d4l3k commented:
Yup

On Sun, Apr 24, 2022, 08:51 Jordan Matelsky ***@***.***>
wrote:

> This is awesome, good to merge @d4l3k <https://github.com/d4l3k>?? Great
> stuff!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/aplbrain/npyjs/pull/26#issuecomment-1107867662>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG56MDDXFCQP7UEOLJIHM3VGVUZDANCNFSM5UEKCULQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

j6k4m8 commented:
Thank you!
